#23 위키 입력 테스트코드 작성

kmmin78 opened the pull request:
개요
위키 입력 테스트코드 작성

AS-IS
위키 로그인 기능만 있는 상태

TO-BE
1. 테스트 유틸클래스 명칭 변경
2. PostServiceTest 클래스 기본 골조 추가
3. PostEntity, Service, Repository 작성
4. Post 관련 UserEntity 작성
5. 본문 100자 이상, 내용 10,000자 이상 시 실패하는 Test Case 작성
6. title 중복 등록 시 exception 발생하는 Test Case 작성

Commit:
#20 위키 입력 테스트코드 작성

주요 내용
1. 테스트 유틸클래스 명칭 변경
2. PostServiceTest 클래스 기본 골조 추가
3. PostEntity, Service, Repository 작성
4. Post 관련 UserEntity 작성
5. 본문 100자 이상, 내용 10,000자 이상 시 실패하는 Test Case 작성
6. title 중복 등록 시 exception 발생하는 Test Case 작성

neropsys commented on app-core/src/main/kotlin/kr/flab/wiki/core/post/repository/PostRepositoryImpl.kt:
```diff
@@ -0,0 +1,13 @@
+
+import kr.flab.wiki.core.post.persistence.Post
+
+class PostRepositoryImpl : PostRepository {
```
별도의 로직 없이 true/false를 리턴한다면 구현체를 구현할 필요 없이 mockito에 PostRepository 인터페이스를  사용하는 것만으로 대체 가능할것 같습니다.
```
mockPostRepository = mock(PostRepository::class.java)
`when`(mockPostRepository.save(post.title)).thenReturn(true)
```

kmmin78 replied:
넵 확실히 service를 테스트하는 수준에선 Repository 구현체가 따로 필요없을 것 같습니다.

neropsys reviewed (dismissed):
PostRepositoryImpl 구현체를 제외하면 문제없습니다

FrancescoJo commented:
PR 본문을 Issue Template 형식에 맞춰 작성해 주세요.

FrancescoJo commented on app-core/src/main/kotlin/kr/flab/wiki/core/post/business/PostService.kt:
```diff
@@ -0,0 +1,7 @@
+import kr.flab.wiki.core.post.persistence.Post
+
+interface PostService {
+    fun writePost(post : Post) : Boolean
```
만약, `Post` 객체의 `id`, `createdAt` 처럼 자동 생성되는 정책을 가진 값이 있을 경우, 외부에서 굳이 값을 넘겨주지 않을 텐데, Service 내에서 처리한 값을 외부에서 알려면 반환값이 어떻게 되어야 할까요?

kmmin78 replied:
결국 입력했던 post객체로 다시 돌려줘야할 것 같습니다.

FrancescoJo commented:
Formatter 한번 적용해 주시기 바랍니다. 공통 Code format 파일은 project root 의 `buildscripts/configuration/codestyle_intellij.xml' 에 있습니다.

FrancescoJo commented on app-core/src/main/kotlin/kr/flab/wiki/core/post/business/PostServiceImpl.kt:
```diff
@@ -0,0 +1,24 @@
+
+class PostServiceImpl(private val postRepository : PostRepository) : PostService {
+
+    private val MAX_TITLE_LENGTH : Int = 100
```
이대로 submit 하시면 detekt 체크에 걸립니다. 이런 식으로 상수로 쓸 값은 `companion object` 아래에 `const val` 로 선언해 주시면 java 의 `final static` 과 같은 효과를 낼 수 있습니다.

kmmin78 replied:
네 코틀린에선 static이 없고, companion object를 사용하는 것 같습니다. 수정해놓겠습니다.

FrancescoJo commented on app-core/src/main/kotlin/kr/flab/wiki/core/post/exception/PostValidationException.kt:
```diff
@@ -0,0 +1,5 @@
+
+import kr.flab.wiki.core.common.exception.ValidationException
+
+class PostValidationException(msg : String) : ValidationException(msg)
```
이런 식의 내부 Business exception 을 구성하는 것은 좋습니다.

FrancescoJo commented on app-core/src/main/kotlin/kr/flab/wiki/core/post/persistence/Post.kt:
```diff
@@ -0,0 +1,11 @@
+import java.time.LocalDateTime
+
+interface Post {
+    val writer : User
```
추후 Post 수정은 어떤 식으로 하실 건가요?

kmmin78 replied:
var로 수정해야될 것 같습니다. 매번 구현체로 형변환해서 수정하는 건 비효율적일 것 같습니다.

FrancescoJo commented on app-core/src/test/kotlin/kr/flab/wiki/core/login/PostServiceTest.kt:
```diff
@@ -0,0 +1,116 @@
+
+//한글 사용을 위해 추가
+@Suppress("NonAsciiCharacters")
+@DisplayName("해당 테스트는 PostService 메서드를 테스트하는 클래스이다.")
```
좋아요

FrancescoJo commented:
코드 퀄리티가 처음보다 눈에 띄게 좋아져서 기쁩니다. 제가 코멘트한 내용들 한번 살펴보시고, 수정 후 다시 말씀해주세요.

Commit:
#20 위키 입력 테스트코드 작성

코드리뷰 피드백 반영
1. PostRepositoryImpl 구현부 삭제
2. writePost return 타입 Post로 변경
3. formatter 적용
4. 상수 companion object로 변경
5. Post 변경 가능한 프로퍼티로 수정

FrancescoJo reviewed (approved):
잘 봤습니다. 고생하셨어요.
